rust/src/array.rs

```diff
@@ -18,447 +18,562 @@
+            .build();
+        BinaryArray::from(array_data)
     }
 }
 
-/// Create a BufferArray<T> from a Vec<Optional<T>> with null handling
-impl<T> From<Vec<Option<T>>> for PrimitiveArray<T>
-where
-    T: ArrowPrimitiveType + 'static,
-{
-    fn from(v: Vec<Option<T>>) -> Self {
-        let mut builder: Builder<T> = Builder::with_capacity(v.len());
-        builder.set_len(v.len());
-        let mut null_count = 0;
-        let mut validity_bitmap = Bitmap::new(v.len());
-        for i in 0..v.len() {
-            match v[i] {
-                Some(value) => builder.set(i, value),
-                None => {
-                    null_count += 1;
-                    validity_bitmap.clear(i);
-                }
-            }
-        }
-        PrimitiveArray::new(builder.finish(), null_count, Some(validity_bitmap))
+impl Array for BinaryArray {
+    fn as_any(&self) -> &Any {
+        self
+    }
+
+    fn data(&self) -> ArrayDataRef {
+        self.data.clone()
+    }
+
+    fn data_ref(&self) -> &ArrayDataRef {
+        &self.data
     }
 }
 
-/// An Array of structs
 pub struct StructArray {
-    len: usize,
-    columns: Vec<Arc<Array>>,
-    null_count: usize,
-    validity_bitmap: Option<Bitmap>,
+    data: ArrayDataRef,
+    boxed_fields: Vec<ArrayRef>,
 }
 
 impl StructArray {
-    pub fn num_columns(&self) -> usize {
-        self.columns.len()
+    /// Returns the field at `pos`.
+    pub fn column(&self, pos: usize) -> &ArrayRef {
+        &self.boxed_fields[pos]
     }
-    pub fn column(&self, i: usize) -> &Arc<Array> {
-        &self.columns[i]
+}
+
+impl From<ArrayDataRef> for StructArray {
+    fn from(data: ArrayDataRef) -> Self {
+        let mut boxed_fields = vec![];
+        for cd in data.child_data() {
+            boxed_fields.push(make_array(cd.clone()));
+        }
+        Self {
+            data: data,
+            boxed_fields: boxed_fields,
+        }
     }
 }
 
 impl Array for StructArray {
-    fn len(&self) -> usize {
-        self.len
-    }
-    fn null_count(&self) -> usize {
-        self.null_count
-    }
-    fn validity_bitmap(&self) -> &Option<Bitmap> {
-        &self.validity_bitmap
-    }
     fn as_any(&self) -> &Any {
         self
     }
+
+    fn data(&self) -> ArrayDataRef {
+        self.data.clone()
+    }
+
+    fn data_ref(&self) -> &ArrayDataRef {
+        &self.data
+    }
 }
 
-/// Create a StructArray from a list of arrays representing the fields of the struct. The fields
-/// must be in the same order as the schema defining the struct.
-impl From<Vec<Arc<Array>>> for StructArray {
-    fn from(data: Vec<Arc<Array>>) -> Self {
-        StructArray {
-            len: data[0].len(),
-            columns: data,
-            null_count: 0,
-            validity_bitmap: None,
+impl From<Vec<ArrayRef>> for StructArray {
+    fn from(v: Vec<ArrayRef>) -> Self {
+        let mut field_types = vec![];
+        let mut count = 0;
+        for arr in &v {
```

Comment:
Why `&v` here while you're using plain `v` in the other `for` loop below?

Reply:
`v` means to consume the unique reference. We cannot consume the reference twice so the first use needs to be `&v` - just read it. We can change the second one to `&v` as well and it should have no difference.